File: 프로그래머스/lv3/42627. 디스크 컨트롤러/디스크 컨트롤러.py
import heapq as hq

def solution(jobs):
    res = []
    time = 0

    jobs.sort()
    job_idx = 0
    job_length = len(jobs)
    
    will_pending = 0
    q = [] #heapq
    
    while True:
        # q에 job 추가
        while job_idx<job_length and jobs[job_idx][0] <= time:
            request_time, cost = jobs[job_idx]
            hq.heappush(q, [cost, request_time])
            job_idx += 1
        # 처리중인 일이 없다면 작업 시작
        if will_pending <= time and q:
            cost, request_time = hq.heappop(q)
            res.append(time-request_time+cost)
            will_pending = time + cost
            time = will_pending
        else:
            time += 1
        
        if job_idx==job_length and not q:
            return sum(res)//len(res)
            
    

        
        
        
        
        
    
    
    
    

File: 프로그래머스/lv3/42627. 디스크 컨트롤러/디스크 컨트롤러.js
function solution(jobs) {
    let res = []
    let time = 0
    let controller = []
    jobs = jobs.sort((a,b) => b[0]-a[0] || b[1]-a[1])

    while (jobs.length || controller.length) {
        while (jobs.length && jobs[jobs.length-1][0] <= time) {
            controller.push(jobs.pop())
        }
        if (controller.length) {
            let [request_time, cost] = controller[0]
            let popIdx = 0
            for (idx = 1; idx < controller.length; idx++) {
                if (controller[idx][1] < cost) {
                    [request_time, cost] = controller[idx]
                    popIdx = idx
                }
            }
            controller.splice(popIdx,1)
            res.push(time-request_time+cost)
            time += cost
        } else {
            time = jobs[jobs.length-1][0]
        }
    }
    return Math.floor(res.reduce((acc,cur)=>acc+cur)/res.length)
}